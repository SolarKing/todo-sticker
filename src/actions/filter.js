// @flow

export type FilterActions =
    | DueDateFilterAction;

export type DueDateFilterAction = {|
    type: 'dueDateFilter',
    filter: string,
|}

export function settingDueDateFilter(dueDateFilter: string): DueDateFilterAction {
    return {
        type: 'dueDateFilter',
        filter: dueDateFilter
    };
}
