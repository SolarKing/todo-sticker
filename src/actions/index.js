// @flow
import type { TodoActions } from './todo';
import type { FilterActions } from './filter';
import type { SortActions } from './sort';

export type Actions =
    | TodoActions
    | FilterActions
    | SortActions;
