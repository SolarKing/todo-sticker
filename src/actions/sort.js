// @flow
import type {SortByType, SortColumn} from '../constant/sort';

export type SortActions =
    | SortByAction
    | SortColumnAction;

export type SortColumnAction = {|
    type: 'sortColumn',
    column: SortColumn
|};

export type SortByAction = {|
    type: 'sortBy',
    sortBy: SortByType
|};

export function settingSortColumn(column: SortColumn): SortColumnAction {
    return {
        type: 'sortColumn',
        column: column
    };
}
