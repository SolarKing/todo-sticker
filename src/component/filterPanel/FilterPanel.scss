.filter-panel {
    position: absolute;
    z-index: 100;
    width: 80%;
    height: auto;
    background-color: #ffffff;
    padding: 10px;
    box-shadow: 0px 0px 3px #888888;
    transition: all .2s ease-in-out;
    transform: scale(1);
    transform-origin: 0% 0%;

    &.close {
        transform: scale(0);
    }

}

.mask {
    top: 0;
    left: 0;
    opacity: 0;
    z-index: 99;
    position: fixed;
    width: 100%;
    height: 100%;
}
