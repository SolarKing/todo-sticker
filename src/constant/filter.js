// @flow
export type DueDateFilter = 'today' | 'all';

export const DueDateFilterEnum = {
    today: 'today',
    all: 'all'
};
