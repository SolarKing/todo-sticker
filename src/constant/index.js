// @flow

const DueDate = {
    today: 'today',
    all: 'all'
};


export const FilterEnum = {
    DueDate
};
