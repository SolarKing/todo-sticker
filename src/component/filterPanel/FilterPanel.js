// @flow
import React from 'react';
import './FilterPanel.scss';
import Translate from '../../class/translate';
import { Radio } from 'antd';
import Classnames from 'classnames';
import {DueDateFilterEnum} from '../../constant/filter';
import type {DueDateFilter} from '../../constant/filter';

type Props = {
    show: boolean,
    handleDueDateFilterChanged: (filter: DueDateFilter) => void,
    onHidden?: () => void
}
type States = {
    dueDateFilter: DueDateFilter,
    show: boolean
}

export default class FilterPanel extends React.Component<Props, States> {

    handleDueDateFilterSelected: Function;
    handleClickOutside: Function;

    static defaultProps = {
        show: false
    }

    state = {
        dueDateFilter: DueDateFilterEnum.all,
        show: false
    }

    constructor(props: Props) {
        super(props);
        this.handleDueDateFilterSelected = this. handleDueDateFilterSelected.bind(this);
        this.handleClickOutside = this. handleClickOutside.bind(this);
    }

    componentWillReceiveProps(nextProps: Props) {
        if (nextProps.show !== this.state.show) {
            this.setState({
                show: nextProps.show
            });
        }
    }

    handleClickOutside() {
        this.setState({
            show: false,
        });
        if (this.props.onHidden) {
            this.props.onHidden();
        }
    }

    handleDueDateFilterSelected(event: any) {
        let seleted: DueDateFilter = event.target.value;
        this.props.handleDueDateFilterChanged(seleted);
    }

    render() {
        let {
            handleDueDateFilterChanged,
            show,
            onHidden,
            ...others
        } = this.props;
        show;
        handleDueDateFilterChanged;
        onHidden;

        let usedClassName = Classnames({
            'filter-panel': true,
            'close': !this.state.show,
        });


        return (
            <div>
                <div className={usedClassName} {...others} >
                    <div className="filter-group" >
                        <p>{Translate.tr('Due Date:')}</p>
                        <Radio.Group onChange={this.handleDueDateFilterSelected} defaultValue={this.state.dueDateFilter}>
                            <Radio.Button value={DueDateFilterEnum.today}>{Translate.tr('Today')}</Radio.Button>
                            <Radio.Button value={DueDateFilterEnum.all}>{Translate.tr('All')}</Radio.Button>
                        </Radio.Group>
                    </div>
                    <hr/>
                    <div className="filter-group" >
                        <p>{Translate.tr('Complete Status:')}</p>
                        <Radio.Group onChange={this.handleDueDateFilterSelected} defaultValue={this.state.dueDateFilter}>
                            <Radio.Button value={DueDateFilterEnum.today}>{Translate.tr('Complete')}</Radio.Button>
                            <Radio.Button value={DueDateFilterEnum.all}>{Translate.tr('Uncompleted')}</Radio.Button>
                        </Radio.Group>
                    </div>
                </div>
                {
                    (this.state.show? (<div onClick={this.handleClickOutside} className="mask" />): null)
                }
            </div>
        );
    }

}
