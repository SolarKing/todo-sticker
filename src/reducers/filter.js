// @flow
import type {FilterState} from '../states/index';
import {DueDateFilterEnum} from '../constant/filter';
import type {FilterActions} from '../actions/filter';

const initialFilter: FilterState = {
    dueDate: DueDateFilterEnum.all,
};

function filter(state: FilterState = initialFilter, action: FilterActions) {
    switch (action.type) {
    case 'dueDateFilter':
        return {
            ...state,
            dueDate: action.filter
        };
    default:
        return  state;
    }
}

export default filter;
