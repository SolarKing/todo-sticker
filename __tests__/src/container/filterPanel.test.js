// @flow
import React from 'react';
import { shallow, type ShallowWrapper } from 'enzyme';
import FilterPanel from '../../../src/container/filterPanel';
import { Radio } from 'antd';
import { DueDateFilterEnum, CompleteStatusFilterEnum } from '../../../src/constant/filter';
import configureStore from 'redux-mock-store';
import { settingDueDateFilter, settingCompleteStatusFilter } from '../../../src/actions/filter';

describe('FilterPanel', () => {
    let wrapper: ShallowWrapper;
    let wrapperContainer: ShallowWrapper;
    let onHide = jest.fn();

    let store = configureStore()({
        filter: {
            dueDate: DueDateFilterEnum.all,
            completeStatus: CompleteStatusFilterEnum.all
        }
    });

    beforeAll(() => {
        wrapperContainer = shallow(
            <FilterPanel
                store={store}
                show={true}
                onHide={onHide}
            />
        );
        wrapper = wrapperContainer.shallow();
    });

    afterEach(() => {
        store.clearActions();
    });

    it('should render correct as container', () => {
        expect(wrapperContainer).toMatchSnapshot();
    });

    it('should render correct only component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should change due date filter', () => {
        const event = {
            target: {
                value: DueDateFilterEnum.today
            }
        };
        wrapper.find(Radio.Group).at(0).simulate('change', event);
        expect(store.getActions()).toEqual([settingDueDateFilter(DueDateFilterEnum.today)]);
    });

    it('should change complete status filter', () => {
        const event = {
            target: {
                value: CompleteStatusFilterEnum.uncomplete
            }
        };
        wrapper.find(Radio.Group).at(1).simulate('change', event);
        expect(store.getActions()).toEqual([settingCompleteStatusFilter(CompleteStatusFilterEnum.uncomplete)]);
    });

    it('should trigger onHidden when click outside', () => {
        wrapper.find('MaskDiv').simulate('click');
        expect(onHide.mock.calls.length).toEqual(1);
    });

});
