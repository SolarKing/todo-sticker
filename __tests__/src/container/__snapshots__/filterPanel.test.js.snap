// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPanel should render correct as container 1`] = `
<FilterPanel
  completeStatus="all"
  dueDate="all"
  onCompleteStatusFilterChanged={[Function]}
  onDueDateFilterChanged={[Function]}
  onHide={[MockFunction]}
  show={true}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`FilterPanel should render correct only component 1`] = `
<div>
  <FilterPanelDiv
    className=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <FilterGroupDiv>
      <p>
        Due Date:
      </p>
      <RadioGroup
        defaultValue="all"
        disabled={false}
        onChange={[Function]}
      >
        <RadioButton
          prefixCls="ant-radio-button"
          value="today"
        >
          Today
        </RadioButton>
        <RadioButton
          prefixCls="ant-radio-button"
          value="all"
        >
          All
        </RadioButton>
      </RadioGroup>
    </FilterGroupDiv>
    <hr />
    <FilterGroupDiv>
      <p>
        Complete Status:
      </p>
      <RadioGroup
        defaultValue="all"
        disabled={false}
        onChange={[Function]}
      >
        <RadioButton
          prefixCls="ant-radio-button"
          value="all"
        >
          All
        </RadioButton>
        <RadioButton
          prefixCls="ant-radio-button"
          value="complete"
        >
          Complete
        </RadioButton>
        <RadioButton
          prefixCls="ant-radio-button"
          value="uncomplete"
        >
          Uncompleted
        </RadioButton>
      </RadioGroup>
    </FilterGroupDiv>
  </FilterPanelDiv>
  <MaskDiv
    onClick={[Function]}
  />
</div>
`;
